src/AggregateFunctions/AggregateFunctionGroupArraySorted.cpp

```diff
@@ -0,0 +1,394 @@
+{
+    heap,
+    sort
+};
+
+constexpr size_t group_array_sorted_sort_strategy_max_elements_threshold = 1000000;
+
+template <typename T, GroupArraySortedStrategy strategy>
+struct GroupArraySortedData
+{
+    using Allocator = MixedAlignedArenaAllocator<alignof(T), 4096>;
+    using Array = PODArray<T, 32, Allocator>;
+
+    static constexpr size_t partial_sort_max_elements_factor = 2;
+
+    static constexpr bool is_value_generic_field = std::is_same_v<T, Field>;
+
+    Array values;
+
+    static bool compare(const T & lhs, const T & rhs)
+    {
+        if constexpr (is_value_generic_field)
+        {
+            return lhs < rhs;
+        }
+        else
+        {
+            return CompareHelper<T>::less(lhs, rhs, -1);
+        }
+    }
+
+    struct Comparator
+    {
+        bool operator()(const T & lhs, const T & rhs)
+        {
+            return compare(lhs, rhs);
+        }
+    };
+
+    ALWAYS_INLINE void heapReplaceTop()
+    {
+        size_t size = values.size();
+        if (size < 2)
+            return;
+
+        size_t child_index = 1;
+
+        if (values.size() > 2 && compare(values[1], values[2]))
+            ++child_index;
+
+        /// Check if we are in order
+        if (compare(values[child_index], values[0]))
+            return;
+
+        size_t current_index = 0;
+        auto current = values[current_index];
+
+        do
+        {
+            /// We are not in heap-order, swap the parent with it's largest child.
+            values[current_index] = values[child_index];
+            current_index = child_index;
+
+            // Recompute the child based off of the updated parent
+            child_index = 2 * child_index + 1;
+
+            if (child_index >= size)
+                break;
+
+            if ((child_index + 1) < size && compare(values[child_index], values[child_index + 1]))
+            {
+                /// Right child exists and is greater than left child.
+                ++child_index;
+            }
+
+            /// Check if we are in order.
+        } while (!compare(values[child_index], current));
+
+        values[current_index] = current;
+    }
+
+    ALWAYS_INLINE void sortAndLimit(size_t max_elements, Arena * arena)
+    {
+        if constexpr (is_value_generic_field)
+            ::sort(values.begin(), values.end(), Comparator());
+        else
+            RadixSort<RadixSortNumTraits<T>>::executeLSD(values.data(), values.size());
+
+        if (values.size() > max_elements)
+            values.resize(max_elements, arena);
+    }
+
+    ALWAYS_INLINE void partialSortAndLimitIfNeeded(size_t max_elements, Arena * arena)
+    {
+        if (values.size() < max_elements * partial_sort_max_elements_factor)
+            return;
+
+        ::nth_element(values.begin(), values.begin() + max_elements, values.end(), Comparator());
+        values.resize(max_elements, arena);
+    }
+
+    ALWAYS_INLINE void addElement(T && element, size_t max_elements, Arena * arena)
```

Comment:
I mentioned (and was ignored) in the previous PR. I feel having addMany could be greatly beneficial here, specially for the Field types. You do a partial sort (or a permutation up to max_element) of the input, then use that to merge into the array. Any thoughts?

Maybe it'd even allow us to forget about having 2 ways of sorting.

Reply:
I suspect that we do not have `addMany` function in `IAggregateFunction`, probably you mean `addBatch` ?
Implementation with heap in practice is much faster, despite having worse asymptotic.
Not sure if sorting is necessary at all.